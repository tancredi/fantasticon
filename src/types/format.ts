import svg2ttf from 'svg2ttf';
import ttf2woff from 'ttf2woff';
import { SvgIcons2FontOptions } from 'svgicons2svgfont';
import { Arguments } from '../types/utils';

type WoffOptions = Arguments<typeof ttf2woff>[1];
type TtfOptions = svg2ttf.FontOptions;
type SvgOptions = Omit<
  SvgIcons2FontOptions,
  'fontName' | 'fontHeight' | 'descent' | 'normalize'
>;

interface JsonOptions {
  indent?: number;
}

interface TsOptions {
  types?: ('enum' | 'constant' | 'charEnum' | 'literalId' | 'literalKey')[];
  singleQuotes?: boolean;
}

export interface FormatOptions {
  woff?: WoffOptions;
  ttf?: TtfOptions;
  svg?: SvgOptions;
  json?: JsonOptions;
  ts?: TsOptions;
}
