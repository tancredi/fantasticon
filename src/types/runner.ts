import { CodepointsMap } from '../utils/codepoints';
import { FontAssetType, OtherAssetType, AssetType, GetIconIdFn } from './misc';
import { FormatOptions } from './format';

export interface RunnerMandatoryOptions {
  inputDir: string;
  outputDir: string;
}

export type RunnerOptionalOptions = {
  name: string;
  fontTypes: FontAssetType[];
  assetTypes: OtherAssetType[];
  formatOptions: FormatOptions;
  pathOptions: { [key in AssetType]?: string };
  codepoints: CodepointsMap;
  fontHeight: number;
  descent: number;
  normalize: boolean;
  round: number;
  selector: string;
  tag: string;
  templates: { [key in OtherAssetType]?: string };
  prefix: string;
  fontsUrl: string;
  getIconId: GetIconIdFn;
  addLigatures: boolean;
};

export type RunnerOptionsInput = RunnerMandatoryOptions &
  Partial<RunnerOptionalOptions>;

export type RunnerOptions = RunnerMandatoryOptions &
  Partial<RunnerOptionalOptions>;
