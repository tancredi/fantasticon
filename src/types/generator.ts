import { AssetsMap } from '../utils/assets';
import { AssetType, OtherAssetType } from './misc';
import { RunnerOptions } from './runner';
import { FormatOptions } from './format';

export type FontGeneratorOptions = RunnerOptions & {
  assets: AssetsMap;
  formatOptions: FormatOptions;
  templates: { [key: string]: string };
};

export type Result = Promise<string | Buffer>;

export type FontGeneratorFn<DependencyT> = (
  options: FontGeneratorOptions,
  dependencyContent: DependencyT extends {} ? DependencyT : null,
  ext?: string
) => Result;

export type FontGenerator<DependencyT = void> = {
  generate: FontGeneratorFn<DependencyT>;
} & (DependencyT extends {} ? { dependsOn: AssetType } : {});
