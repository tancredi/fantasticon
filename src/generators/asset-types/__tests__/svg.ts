import * as _SVGIcons2SVGFontStream from 'svgicons2svgfont';
import { FontAssetType } from '../../../types/misc';
import { FontGeneratorOptions } from '../../../types/generator';
import svgGen from '../svg';

const SVGIcons2SVGFontStream = _SVGIcons2SVGFontStream as unknown as jest.Mock<
  typeof _SVGIcons2SVGFontStream
>;

jest.mock('fs', () => ({
  createReadStream: (filepath: string) => ({
    content: `content->${filepath}`
  })
}));

jest.mock('svgicons2svgfont', () => {
  const { EventEmitter } = require('events');

  class MockStream {
    public events = new EventEmitter();
    public content = '';

    public write(chunk: any) {
      this.events.emit(
        'data',
        Buffer.from(
          `processed->${chunk.content}|${JSON.stringify(chunk.metadata)}$`
        )
      );
      return this;
    }

    public on(event: string, callback: () => void) {
      this.events.on(event, callback);
      return this;
    }

    public end() {
      this.events.emit('end');
      return this;
    }
  }

  return jest.fn(() => new MockStream());
});

const mockOptions = (svgOptions = { __mock: 'options__' } as any) =>
  ({
    name: 'foo',
    fontHeight: 1,
    descent: 2,
    normalize: false,
    formatOptions: { [FontAssetType.SVG]: svgOptions },
    codepoints: { foo: 1, bar: 1 },
    assets: {
      foo: { id: 'foo', absolutePath: '/root/foo.svg' },
      bar: { id: 'bar', absolutePath: '/root/bar.svg' }
    }
  } as unknown as FontGeneratorOptions);

const mockOptionsWithLigatures = (
  svgOptions = { __mock: 'options__' } as any
) =>
  ({
    name: 'foo',
    fontHeight: 1,
    descent: 2,
    normalize: false,
    formatOptions: { [FontAssetType.SVG]: svgOptions },
    codepoints: { foo: 1, bar: 1 },
    assets: {
      foo: { id: 'foo', absolutePath: '/root/foo.svg' },
      bar: { id: 'bar', absolutePath: '/root/bar.svg' }
    },
    addLigatures: true
  } as unknown as FontGeneratorOptions);

describe('`SVG` font generator', () => {
  beforeEach(() => {
    SVGIcons2SVGFontStream.mockClear();
  });

  test('resolves with the result of the completed `SVGIcons2SVGFontStream`', async () => {
    const result = await svgGen.generate(mockOptions(), null);

    expect(SVGIcons2SVGFontStream).toHaveBeenCalledTimes(1);
    expect(SVGIcons2SVGFontStream).toHaveBeenCalledWith({
      descent: 2,
      fontHeight: 1,
      fontName: 'foo',
      log: expect.any(Function),
      normalize: false,
      __mock: 'options__'
    });

    expect(result).toMatchSnapshot();
  });

  test('passes correctly format options to `SVGIcons2SVGFontStream`', async () => {
    const log = () => null;
    const formatOptions = { descent: 5, fontHeight: 6, log };
    const result = await svgGen.generate(mockOptions(formatOptions), null);

    expect(result).toMatchSnapshot();

    expect(SVGIcons2SVGFontStream).toHaveBeenCalledTimes(1);
    expect(SVGIcons2SVGFontStream).toHaveBeenCalledWith({
      descent: 5,
      fontHeight: 6,
      fontName: 'foo',
      log,
      normalize: false
    });
  });

  test('resolves with the result of the completed `SVGIcons2SVGFontStream` with ligatures', async () => {
    const result = await svgGen.generate(mockOptionsWithLigatures(), null);

    expect(SVGIcons2SVGFontStream).toHaveBeenCalledTimes(1);
    expect(SVGIcons2SVGFontStream).toHaveBeenCalledWith({
      descent: 2,
      fontHeight: 1,
      fontName: 'foo',
      log: expect.any(Function),
      normalize: false,
      __mock: 'options__'
    });

    expect(result).toMatchSnapshot();
  });

  test('passes correctly format options to `SVGIcons2SVGFontStream` with ligatures', async () => {
    const log = () => null;
    const formatOptions = { descent: 5, fontHeight: 6, log };
    const result = await svgGen.generate(
      mockOptionsWithLigatures(formatOptions),
      null
    );

    expect(result).toMatchSnapshot();

    expect(SVGIcons2SVGFontStream).toHaveBeenCalledTimes(1);
    expect(SVGIcons2SVGFontStream).toHaveBeenCalledWith({
      descent: 5,
      fontHeight: 6,
      fontName: 'foo',
      log,
      normalize: false
    });
  });
});
