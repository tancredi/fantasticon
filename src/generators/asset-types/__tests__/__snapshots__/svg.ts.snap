// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`SVG\` font generator passes correctly format options to \`SVGIcons2SVGFontStream\` 1`] = `"processed->content->/root/foo.svg|{\\"name\\":\\"foo\\",\\"unicode\\":[\\"\\\\u0001\\"]}$processed->content->/root/bar.svg|{\\"name\\":\\"bar\\",\\"unicode\\":[\\"\\\\u0001\\"]}$"`;

exports[`\`SVG\` font generator passes correctly format options to \`SVGIcons2SVGFontStream\` with ligatures 1`] = `"processed->content->/root/foo.svg|{\\"name\\":\\"foo\\",\\"unicode\\":[\\"\\\\u0001\\",\\"foo\\"]}$processed->content->/root/bar.svg|{\\"name\\":\\"bar\\",\\"unicode\\":[\\"\\\\u0001\\",\\"bar\\"]}$"`;

exports[`\`SVG\` font generator resolves with the result of the completed \`SVGIcons2SVGFontStream\` 1`] = `"processed->content->/root/foo.svg|{\\"name\\":\\"foo\\",\\"unicode\\":[\\"\\\\u0001\\"]}$processed->content->/root/bar.svg|{\\"name\\":\\"bar\\",\\"unicode\\":[\\"\\\\u0001\\"]}$"`;

exports[`\`SVG\` font generator resolves with the result of the completed \`SVGIcons2SVGFontStream\` with ligatures 1`] = `"processed->content->/root/foo.svg|{\\"name\\":\\"foo\\",\\"unicode\\":[\\"\\\\u0001\\",\\"foo\\"]}$processed->content->/root/bar.svg|{\\"name\\":\\"bar\\",\\"unicode\\":[\\"\\\\u0001\\",\\"bar\\"]}$"`;
